netket/dynamics.py

```diff
@@ -1 +1,362 @@
+
+    self._obs_samples_valid = False
+    self._sampler.reset()
+    self._sampler.generate_samples(self._n_discard)
+    self._samples = self._sampler.generate_samples(
+        self._n_samples_node, samples=self._samples
+    )
+    self._lloc, self._loss_stats = self._get_mc_superop_stats(self._lind)
+    self._loss1_stats = _statistics(self._lloc.T)
+    lloc = self._lloc - _mean(self._lloc)
+    samples_r = self._samples.reshape((-1, self._samples.shape[-1]))
+    lloc_r = lloc.reshape(-1, 1)
+    self._grads, self._jac = self._machine.vector_jacobian_prod(
+        samples_r, lloc_r / self._n_samples, self._grads, return_jacobian=True
+    )
+    self._grads = tree_map(_sum_inplace, self._grads)
+    self._dp = self._sr.compute_update(self._jac, self._grads, self._dp)
+    return self._dp
+
+
+def create_timevo(op, sampler, *args, julia=False, **kwargs):
+    """
+    Constructs the correct driver given the forward-time evolution
+    generator `op`, the sampler `sampler` and other arguments.
+
+        This function also constructs the function computing the function
+        (or derivative) defining the ode, and returns it.
+
+        If `julia=False`, then the function has signature fun(time, weights),
+        otherwise it has signature fun(weights, params, time) and the second
+        argument is ignored.
+        This option is thought to be used with integrators from
+        diffeq.py
+    """
+
+    machine = sampler._machine
+    if not isinstance(machine, AbstractDensityMatrix):
+        driver = Vmc(op, sampler, *args, optimizer=None, **kwargs)
+    else:
+        driver = SteadyState(op, sampler, *args, optimizer=None, **kwargs)
+
+    flatten = machine.numpy_flatten
+    unflatten = lambda w: machine.numpy_unflatten(w, machine.parameters)
+
+    def _fun(t, w):
+        machine.parameters = unflatten(w)
+        dp = _time_evo(driver)
+        return flatten(dp)
+
+    if julia:
+
+        def fun(u, p, t):
+            _fun(t, w)
+
+    else:
+        fun = _fun
+
+    return driver, fun
+
+
+METHODS = METHODS = {
+    "RK23": scipy.integrate.RK23,
+    "RK45": scipy.integrate.RK45,
+    "DOP853": scipy.integrate.DOP853,
+    "RADAU": scipy.integrate.Radau,
+    "BDF": scipy.integrate.BDF,
+    "LSODA": scipy.integrate.LSODA,
+    "EULER": EulerSolver,
+}
+
+
+class TimeEvolution(AbstractVariationalDriver):
+    """
+    Energy minimization using Variational Monte Carlo (VMC).
+    """
+
+    def __init__(
+        self,
+        operator,
+        sampler,
+        sr,
+        *args,
+        **kwargs,
+    ):
+        driver, fun = create_timevo(operator, sampler, *args, sr=sr, **kwargs)
+
+        super(TimeEvolution, self).__init__(
+            sampler.machine, None, minimized_quantity_name=driver._loss_name
+        )
+        self._driver = driver
+        self._fun = fun
+
+        self._solver = None
+        self._state = None
+
+    def solver(self, METHOD, tspan=0.0, dt=None, adaptive=False, **kwargs):
```

Comment:
I concur with @gcarleo that it would be more consistent with the rest of the interface to pass the solver as an argument to `TimeEvolution.__init__`. instead of separately having to call `.solver` after creating the driver.

Maybe we can support an interface where the solver passed to `TimeEvolution.__init__` is a function taking the ode function as an argument and returning the initialized solver. This would get us around having to wrap each scipy solver in a more complicated fashion. If I am not mistaken, in terms of the current API that would mean something like
```python
solver = lambda ode_fun: METHODS["RK23"](ode_fun, t0, y0, ...)
# The right-hand side here can be packaged as `nk.dynamics.ode_solver` or similar, so that
#     solver = nk.dynamics.ode_solver("RK23", tspan, ...)
# with the current `tevo.solver` args would be equivalent.
tevo = TimeEvolution(..., ode_solver=solver)
```
and then within the initializer one could do
```python
self._integrator = ode_solver(self._fun)
```
So essentially we extract `TimeEvolution.solver` as external method with delayed passing of the ODE function.